src/hotspot/share/jfr/periodic/sampling/jfrThreadSampling.cpp

```diff
@@ -306,6 +351,32 @@ static void drain_enqueued_requests(const JfrTicks& now, JfrThreadLocal* tl, Jav
   assert(!tl->has_enqueued_requests(), "invariant");
 }
 
+
+static void drain_all_enqueued_requests(const JfrTicks& now, JfrThreadLocal* tl, JavaThread* jt, Thread* current) {
+  assert(tl != nullptr, "invariant");
+  assert(jt != nullptr, "invariant");
+  assert(current != nullptr, "invariant");
+  drain_enqueued_requests(now, tl, jt, current);
+#ifdef LINUX
+  if (tl->has_cpu_time_jfr_requests()) {
+    JfrTicks now = JfrTicks::now();
+    tl->acquire_cpu_time_jfr_dequeue_lock();
+    JfrCPUTimeTraceQueue& queue = tl->cpu_time_jfr_queue();
+    for (u4 i = 0; i < queue.size(); i++) {
+      record_cpu_time_thread(queue.at(i), now, tl, jt, current);
+    }
+    queue.set_size(0);
+    assert(queue.is_empty(), "invariant");
+    tl->set_has_cpu_time_jfr_requests(false);
+    if (queue.lost_samples() > 0) {
+      JfrCPUTimeThreadSampling::send_lost_event( now, JfrThreadLocal::thread_id(jt), queue.get_and_reset_lost_samples());
+    }
+    queue.clear();
+    tl->release_cpu_time_jfr_queue_lock();
```

Comment:
Is this releasing a different lock from the one acquired? "dequeue" lock vs "queue" lock?

Reply:
These are not really different locks, as the lock has four different states. So it just changes the lock state from DEQUEUE to UNLOCKED.